google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/gaxx/grpc/ChannelPoolHealthChecker.java

```diff
@@ -0,0 +1,225 @@
+      success = true;
+    } catch (Exception e) {
+      success = false;
+    }
+    addProbeResult(entry, ProbeResult.create(startTime, success));
+  }
+
+  @VisibleForTesting
+  void addProbeResult(Entry entry, ProbeResult result) {
+    entry.probeHistory.add(result);
+    if (result.isSuccessful()) {
+      entry.successfulProbesInWindow.incrementAndGet();
+    } else {
+      entry.failedProbesInWindow.incrementAndGet();
+    }
+  }
+
+  @VisibleForTesting
+  void pruneHistory(Entry entry) {
+    Instant windowStart = clock.instant().minus(WINDOW_DURATION);
+    while (!entry.probeHistory.isEmpty()
+        && entry.probeHistory.peek().startTime().isBefore(windowStart)) {
+      ProbeResult removedResult = entry.probeHistory.poll();
+      if (removedResult.isSuccessful()) {
+        entry.successfulProbesInWindow.decrementAndGet();
+      } else {
+        entry.failedProbesInWindow.decrementAndGet();
+      }
+    }
+  }
+
+  /** Checks if a single entry is currently healthy based on its probe history. */
+  @VisibleForTesting
+  boolean isEntryHealthy(Entry entry) {
+    int failedProbes = entry.failedProbesInWindow.get();
+    int totalProbes = failedProbes + entry.successfulProbesInWindow.get();
+
+    if (totalProbes < MIN_PROBES_FOR_EVALUATION) {
+      return true; // Not enough data, assume healthy.
+    }
+
+    double failureRate = ((double) failedProbes / totalProbes) * 100.0;
+    return failureRate < SINGLE_CHANNEL_FAILURE_PERCENT_THRESHOLD;
+  }
+
+  /**
+   * Finds a channel that is an outlier in terms of health.
+   *
+   * @return the entry to be evicted. Returns null if nothing to evict.
+   */
+  @Nullable
+  @VisibleForTesting
+  Entry findOutlierEntry() {
+    List<Entry> unhealthyEntries =
+        this.entrySupplier.get().stream()
+            .peek(this::pruneHistory)
+            .filter(entry -> !isEntryHealthy(entry))
+            .collect(Collectors.toList());
+
+    int poolSize = this.entrySupplier.get().size();
+    if (unhealthyEntries.isEmpty() || poolSize == 0) {
+      return null;
+    }
+
+    // If more than CIRCUITBREAKER_PERCENT of channels are unhealthy we won't evict
+    double unhealthyPercent = (double) unhealthyEntries.size() / poolSize * 100.0;
+    if (unhealthyPercent >= POOLWIDE_BAD_CHANNEL_CIRCUITBREAKER_PERCENT) {
+      return null;
+    }
+
+    return unhealthyEntries.stream()
+        .max(Comparator.comparingInt(entry -> entry.failedProbesInWindow.get()))
+        .orElse(null);
+  }
+
+  /** Periodically detects and removes outlier channels from the pool. */
+  @VisibleForTesting
+  void detectAndRemoveOutlierEntries() {
+    if (clock.instant().isBefore(lastEviction.plus(MIN_EVICTION_INTERVAL))) {
+      // Primitive but effective rate-limiting.
+      return;
+    }
+    Entry outlier = findOutlierEntry();
+    if (outlier != null) {
+      this.lastEviction = clock.instant();
+      outlier.getManagedChannel().enterIdle();
```

Comment:
do you need to clear out the old results now that the underlying socket has changed?